Challenge 2 - Store data in the database

ccjjmartin opened the pull request:
## This PR makes a couple of changes:
- It creates a table named policy_review to store our form information in
- It sends an insert query during the form submission to record our data
- It sends select queries to check for existing data
## Steps to test your new table in the database:
- Follow the instructions from pull request 1 for a fresh install
- Verify that the inope_policy_review module was enabled (and installed)
- Check your database for a new table named policy_review with columns:
  - policy_review_id (primary key)
  - vid
  - nid
  - uid
  - note_status
  - note_submitted
  - note
- Go to /upgrade.php for existing websites (verify that you have update 7100 to perform)
- Check your database for a new table named policy_review with columns mentioned above
- NOTE: If you are uninstalling the module to test this make sure you uninstall feature_inope_policy also

If your vm is named inope and you are using drupal vm then you can:

Copy and paste to uninstall necessary modules:

```
drush @drupalvm.inope.drupalvm.dev dis feature_inope_policy -y; drush @drupalvm.inope.drupalvm.dev dis inope_policy_review -y; drush @drupalvm.inope.drupalvm.dev pmu feature_inope_policy -y; drush @drupalvm.inope.drupalvm.dev pmu inope_policy_review -y;
```

Copy and paste to reinstall necessary modules:

```
drush @drupalvm.inope.drupalvm.dev en feature_inope_policy -y; drush @drupalvm.inope.drupalvm.dev en inope_policy_review -y;
```
## Steps to test functionality - Setup:
- Create two users (admin and staff)
- Log in as admin
- Create a node of the policy type (employee handbook or vacation policy)
## Steps to test functionality - Per User
- View the new node, then click on the "Review of {node_title}" tab
- Submit approved status
- Verify that you don't see the form anymore and see a message thanking you for your input
- Once this is verified open Sequel Pro and truncate the table
  - This is so we can submit input again as the same user
- Submit a changes requested status with some text describing you desired change
- Verify that you don't see the form anymore and see a message thanking you for your input
- Repeat these steps for the staff user
## Question:
- My current understanding is that although I declared the foreign keys in the hook_schema() they are not actually used in MySQL.  I read somewhere that Drupal doesn't take advantage of this yet and it is more of a documentation thing than a functional thing.  Is this still the case or is my information outdated?

ccjjmartin committed:
Challenge 2 - Store data in the database

Custom table using the Schema API
db_select, db_insert, db_update were used to store information

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.install:
```diff
@@ -0,0 +1,93 @@
+ * Note that we only need to do this while upgrading as drupal_install_schema
+ * is called by default in hook_install().
+ */
+function inope_policy_review_update_7000() {
```
Trick for new update hooks - start them at 7100.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -75,11 +75,52 @@ function inope_policy_review_permission() {
 /**
+ * Implements hook_form_alter().
+ */
+function inope_policy_review_form_alter(&$form, &$form_state, $form_id) {
```
Why are you doing this in an alter hook and not in the page callback or the form callback?

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -75,11 +75,52 @@ function inope_policy_review_permission() {
+
+    // Query the policy review table for any records related to the current
+    // node, unapproved notes, and are from the current user.
+    $results = db_select('policy_review', 'pr')
```
The trick about db_selects is that -- unless your query is truly dynamic, that is, the clauses and conditions are assembled at run time and vary tremendously based on interaction or context -- it is slower than db_query. I think I mentioned that in our DB stuff, but if I didn't there it is.

So -- use dbquery for selects only -- (unless your query is very dynamic).   You can find a ton of threads from Larry Garfield on this subject. I used to insist that db_select be used -- but its slow by comparison and that opinion was wrong headed.

Use the database API for everything else.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -75,11 +75,52 @@ function inope_policy_review_permission() {
+
+    // If the active user has submitted something and is not an administrator
+    // clear the form items we don't want to show and show a message.
+    if (!empty($results) && !in_array('administrator', $account->roles)) {
```
You have a couple options:
1) You could see if the form needs to be rendered at all in the page callback
2) You could conditionally not add these items in the first place based on if this needs to be rendered in the first place -- that's probably prone to caching problems though.  That may be why you were using alters. hmm, - I guess I'd probably come down on the side of 1 then -- because if work doesn't need to be done because of a business rule -- a good rule of thumb is to not do the work.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -152,22 +193,57 @@ function inope_policy_review_form_submit($form, &$form_state) {
-    // TODO: put this into the database.
+
+    // Only administrators should be able to approve notes.
+    if (in_array('administrator', $account->roles)) {
```
Is there some weird phrasing in the business problem that you pulled this from?  If so -- sorry.  Both staff and administrators should be able to say that they approve of the policy.  THey are not approving it for the company - but just saying that they approve of the policy for themselves.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -152,22 +193,57 @@ function inope_policy_review_form_submit($form, &$form_state) {
-    drupal_set_message(t('Policy approved for use'));
-    // TODO: put this into the database.
+
+    // Only administrators should be able to approve notes.
```
See our discussion in chat -- but basically we need to allow staff and admins to answer this in the positive.
The client has not asked for the administrators to change the answers of all staff - so this won't be needed.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -152,22 +193,57 @@ function inope_policy_review_form_submit($form, &$form_state) {
-    drupal_set_message(t('Node ID : @nodeid', array('@nodeid' => $node->nid)));
-    drupal_set_message(t('Node VID : @nodevid', array('@nodevid' => $node->vid)));
-    // TODO: put this into the database.
+    drupal_set_message($text);
```
probably want to validate this text before you repost it. Using user submitted material straight up is probably not a great idea -- even in an intranet -- they may copy something or something else intentionally so -- while I think we can generally assume that form monkeying is not in place, I think this should probably be scrubbed

ccjjmartin committed:
Challenge 2 - Review Comments

Fix queries, change approval functionality, remove form alter,
change update hook number

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -55,10 +55,31 @@ function inope_policy_review_tab_title_callback($node) {
+
+  // TODO: Ask client if they want previously submitted approval to be allowed
+  // to be overwritten, this is the current functionality.
+  $results = inope_policy_review_select_query($node, $account, 0);
```
Why do you have status as a part of this function and query?

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -177,3 +210,80 @@ function inope_policy_review_form_submit($form, &$form_state) {
+ * Query the policy review table for any records related to the current
+ * node, note status, and are from the current user.
+ */
+function inope_policy_review_select_query($node, $account, $status) {
```
By using status here - It sort of then says -- that only if a user has submitted that they approve the policy will we hide the form. If they have answered 1 and provided text then -- it will return results and then we will build the form -- which isn't what we want in either this challenge or challenge 3.

So from the above you pass in 0 as a status as a hard coded value meaning that - if they selected 1 as their initial response - or that they do have comments this will return results and rebuild the form.

What we actually need to evaluate is -- did this user submit anything since the most recent timestamp on the most recent vid.  If they have a timestamp that is >= to that timestamp then they don't need to see the form.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -55,10 +55,31 @@ function inope_policy_review_tab_title_callback($node) {
+
+  // TODO: When the form is rebuilt query above this doesn't pick up the new
+  // entry.
+  // If the active user has previously submitted something only display a
```
So this would actually only display this message if they replied with a status 0 given the above function call and query.

thebruce replied:
Unless the problem wasn't clear, which is a possibility, this is telling me that this branch wasn't functionally tested. That is a real no-no. So I am hoping it was the former, but if it was the later, each PR needs to be functionally tested by you for the business requirements. It's kind of assumed that this has happened -- so that a proper code review can take place.  Then a thorough functional code review will happen on the code reviewers side.  Again that second one assumes that you've already had a thorough functional on your side.

thebruce commented:
I'll do a functional after you fix the query and function.

ccjjmartin committed:
Challenge 2 - Fix select query and don't rebuild the form

ccjjmartin committed:
Challenge 2 - phpcs warnings and errors

ccjjmartin committed:
Challenge 2 - Change the base of our SELECT query to timestamp

The original query used status but it should use timestamp.

ccjjmartin committed:
Challenge 2 - Fix insert query and other calls to the select query

ccjjmartin committed:
Challenge 2 - Fix index so that it references the right fields

thebruce commented:
One problem is that if you create another node of a different type (page for example) you see this tab on that form.

We probably need to add in a conditional to make sure that this tab only appears when it is a certain kind of node.

thebruce commented:
The status that is stored in the database is exactly the opposite of the key value in the form for some reason -- when I submitted an approve the status that was recorded was a 1, when I submitted a changes it was recorded as a 0.

thebruce commented:
Table is recorded, users can submit on multiple nodes, status is weird for some reason.

Fix the status and I'll do a code review -- hits the other items (except for the bug where it shows up on all nodes regardless of type).

You might want to attack that via an access callback -- maybe? maybe not?

thebruce commented:
Update -- in the next challenge the status being recorded incorrectly seems to be fixed -- so I will just assume you got it here and do a code review.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -149,31 +169,117 @@ function inope_policy_review_form_submit($form, &$form_state) {
+
+  // Select query to get the latest node revision timestamp.
+  $node_revision = db_query("
+  SELECT vid, timestamp FROM {node_revision} WHERE vid = :vid", array(
```
I think we forgot to turn on revisions for the policy node -- so we should probably do that via features just so this will work properly.

thebruce replied:
changed on the node table should give you the latest update timestamp, i think?

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -149,31 +169,117 @@ function inope_policy_review_form_submit($form, &$form_state) {
+}
+
+/**
+ * Update query if the user wants to update a record.
```
I think this is probably still updating history -- not sure there is any need for an update -- we still want to observe that they originally approved or denied the policy, you can stick with inserts that preserve the entire history.

thebruce replied:
Ah I think there was yet another misunderstanding -- that if we had had more time we could have cleared up. It isn't important that a staff member always be able to update a response they had but rather -- they should be able to update only their current response. I now see why you chose an update here.

thebruce commented:
A few changes.  I wouldn't describe this round of review as thorough -- but those things stuck out.

I appreciate the commenting - it is helping me get the gist quickly, thank you for that.

ccjjmartin committed:
Challenge 2 - Only display on policy review node, feature update, query

Fix the policy review page from appearing on other node types.
Update features for defaulting to creating a new node revision.
Improve select query to a single db_query using $node->>changed.

ccjjmartin commented:
I was thinking about if I should use an insert or an update.  During our planning phase of the table we didn't mention a column (field) for policy review note revision so I came to the conclusion that an update was probably appropriate while preserving as much history as possible but allowing the user to fix things like typos or change their response type.

I am glad you brought up the tab appearing on other node types, that is a big functional error and it has been fixed.

I am glad you saw that the status was wrong and that you gave the go ahead for allowing the fix to come through in the next challenge.

ccjjmartin committed:
Challenge 2 - phpcs lint changes

thebruce commented:
also in answer to your question -- FK are not respected/used by Drupal - they rely on code only to interpret and write correct values, that may change someday, but not in 7.  The main reason is they are trying to have all of the logic in php not share the logic in the database.  Just get me and mike started on this and watch how we disagree, it doesn't matter though because in the end, no one actually believes my side is correct, and for an open source project that needs to install on a variety of hosts with a variety of backends, they are "probably" right.

thebruce commented:
If your satisfied with this you can merge it -- if its the last PR that's going to be merged in though - i would fix the 0/1 problem in this PR.
